homeassistant/components/zha/core/gateway.py

```diff
@@ -143,23 +145,27 @@ def device_left(self, device):
         """Handle device leaving the network."""
         pass
 
-    async def _async_remove_device(self, device):
+    async def _async_remove_device(self, device, entity_refs):
+        if entity_refs is not None:
+            remove_tasks = []
+            for entity_ref in entity_refs:
+                remove_tasks.append(entity_ref.entity.async_remove())
+            await asyncio.gather(*remove_tasks)
         ha_device_registry = await get_dev_reg(self._hass)
         reg_device = ha_device_registry.async_get_device(
             {(DOMAIN, str(device.ieee))}, set())
-        ha_device_registry.async_remove_device(reg_device.id)
+        if reg_device is not None:
+            ha_device_registry.async_remove_device(reg_device.id)
 
     def device_removed(self, device):
         """Handle device being removed from the network."""
         zha_device = self._devices.pop(device.ieee, None)
-        self._device_registry.pop(device.ieee, None)
+        entity_refs = self._device_registry.pop(device.ieee, None)
         if zha_device is not None:
             device_info = async_get_device_info(self._hass, zha_device)
             zha_device.async_unsub_dispatcher()
-            asyncio.ensure_future(self._async_remove_device(zha_device))
-            async_dispatcher_send(
-                self._hass,
-                "{}_{}".format(SIGNAL_REMOVE, str(zha_device.ieee))
+            asyncio.ensure_future(
```

Comment:
Can we use `hass.async_create_task` instead?

Reply:
On a previous PR I was asked to avoid that IIRC. I’ll see if I can find it.

Reply:
Wasn’t in a PR it was in a conversation on discord. I was told that create task messes with stack traces... I’d like to know the preference here though.

Reply:
We prefer to await coroutines directly and catch exceptions early inside scheduled jobs. See https://github.com/home-assistant/home-assistant/issues/15446.

I'm not aware that `asyncio.ensure_future` is better in this regard than `loop.create_task`.